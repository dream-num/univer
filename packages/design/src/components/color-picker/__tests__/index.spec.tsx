/**
 * Copyright 2023-present DreamNum Co., Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { ComponentProps } from 'react';
import { cleanup, render } from '@testing-library/react';
import { useState } from 'react';
import { afterEach, describe, expect, it, vi } from 'vitest';
import { ColorInput } from '../ColorInput';
import { ColorPicker } from '../ColorPicker';
import { ColorSpectrum } from '../ColorSpectrum';
import { HueSlider } from '../HueSlider';
import '@testing-library/jest-dom/vitest';

afterEach(cleanup);

describe('ColorPicker', () => {
    it('should render correctly', () => {
        const { container } = render(<ColorPicker />);

        expect(container).toMatchSnapshot();
    });

    it('should render with value', () => {
        const { container } = render(<ColorPicker value="#FF0000" />);

        expect(container).toMatchSnapshot();
    });

    it('should call onChange when color changes', () => {
        const handleChange = vi.fn();
        const { container } = render(<ColorPicker onChange={handleChange} />);
        const presetBtn = container.querySelector('[data-u-comp="color-picker-presets"] button');
        if (presetBtn) {
            presetBtn.dispatchEvent(new MouseEvent('click', { bubbles: true }));
            expect(handleChange).toHaveBeenCalled();
        }
    });

    it('should support format hex', () => {
        const { container } = render(<ColorPicker format="hex" value="#00FF00" />);
        expect(container).toMatchSnapshot();
    });

    it('should open dialog when more is clicked', () => {
        const { container } = render(<ColorPicker />);
        const moreLink = Array.from(container.querySelectorAll('a')).find((a) => a.textContent?.includes('更多') || a.textContent?.toLowerCase().includes('more'));
        if (moreLink) {
            moreLink.dispatchEvent(new MouseEvent('click', { bubbles: true }));
            expect(document.body.innerHTML).toContain('univer-grid univer-w-64 univer-gap-2');
        }
    });

    it('should call onChange when rgb input changes in dialog', () => {
        const handleChange = vi.fn();
        const { container } = render(<ColorPicker onChange={handleChange} />);
        const moreLink = Array.from(container.querySelectorAll('a')).find((a) => a.textContent?.includes('更多') || a.textContent?.toLowerCase().includes('more'));
        if (moreLink) {
            moreLink.dispatchEvent(new MouseEvent('click', { bubbles: true }));
            const rgbInputs = Array.from(document.querySelectorAll('input')).filter((input) => input.maxLength === 3) as HTMLInputElement[];
            if (rgbInputs.length === 3) {
                rgbInputs[0].value = '1';
                rgbInputs[0].dispatchEvent(new Event('input', { bubbles: true }));
                rgbInputs[1].value = '2';
                rgbInputs[1].dispatchEvent(new Event('input', { bubbles: true }));
                rgbInputs[2].value = '3';
                rgbInputs[2].dispatchEvent(new Event('input', { bubbles: true }));
                const confirmBtn = Array.from(document.querySelectorAll('button')).find((btn) => btn.textContent?.includes('确定') || btn.textContent?.toLowerCase().includes('confirm'));
                if (confirmBtn) {
                    confirmBtn.dispatchEvent(new MouseEvent('click', { bubbles: true }));
                    expect(handleChange).toHaveBeenCalled();
                }
            }
        }
    });
});

describe('HueSlider', () => {
    const onChange = vi.fn();

    function TestComponent(props: { onChange: typeof onChange } & Omit<ComponentProps<typeof HueSlider>, 'hsv' | 'onChange'>) {
        const { onChange, ...restProps } = props;
        const [hsv, setHsv] = useState<[number, number, number]>([0, 100, 100]);
        function handleChange(h: number, s: number, v: number) {
            setHsv([h, s, v]);
            onChange([h, s, v]);
        }

        return <HueSlider hsv={hsv} onChange={handleChange} {...restProps} />;
    }

    it('should call onChange when hue changes by pointer', () => {
        const { container } = render(<TestComponent onChange={onChange} />);
        const slider = container.querySelector('[data-u-comp="color-picker-hue-slider"] > div');
        if (slider) {
            slider.dispatchEvent(new PointerEvent('pointerdown', { bubbles: true, clientX: 50 }));
            expect(onChange).toHaveBeenCalled();
        }
    });
});

describe('ColorSpectrum', () => {
    const onChange = vi.fn();

    function TestComponent(props: { onChange: typeof onChange } & Omit<ComponentProps<typeof ColorSpectrum>, 'hsv' | 'onChange'>) {
        const { onChange, ...restProps } = props;
        const [hsv, setHsv] = useState<[number, number, number]>([0, 100, 100]);
        function handleChange(h: number, s: number, v: number) {
            setHsv([h, s, v]);
            onChange([h, s, v]);
        }

        return <ColorSpectrum hsv={hsv} onChange={handleChange} {...restProps} />;
    }

    it('should call onChange when color changes by pointer', () => {
        const { container } = render(<TestComponent onChange={onChange} />);
        const spectrum = container.querySelector('[data-u-comp="color-picker-spectrum" > canvas]');
        if (spectrum) {
            spectrum.dispatchEvent(new PointerEvent('pointerdown', { bubbles: true, clientX: 50 }));
            expect(onChange).toHaveBeenCalled();
        }
    });
});

describe('ColorInput', () => {
    const onChange = vi.fn();

    function TestComponent(props: { onChange: typeof onChange } & Omit<ComponentProps<typeof ColorInput>, 'hsv' | 'onChange'>) {
        const { onChange, ...restProps } = props;
        const [value, setValue] = useState<[number, number, number]>([0, 100, 100]);

        function handleChange(h: number, s: number, v: number) {
            setValue([h, s, v]);
            onChange([h, s, v]);
        }

        return (
            <ColorInput hsv={value} onChange={handleChange} {...restProps} />
        );
    }

    it('should call onChange when color input changes', () => {
        const { container } = render(<TestComponent onChange={onChange} />);
        const inputs = container.querySelectorAll('input[type="text"]') as unknown as HTMLInputElement[];

        const [rInput, gInput, bInput] = inputs;
        if (rInput && gInput && bInput) {
            rInput.value = '#FF0000';
            gInput.value = '#00FF00';
            bInput.value = '#0000FF';
            rInput.dispatchEvent(new Event('input', { bubbles: true }));
            gInput.dispatchEvent(new Event('input', { bubbles: true }));
            bInput.dispatchEvent(new Event('input', { bubbles: true }));
            expect(onChange).toHaveBeenCalledWith([0, 100, 100]); // Assuming #FF0000 corresponds to HSV [0, 100, 100]
        }
    });
});
