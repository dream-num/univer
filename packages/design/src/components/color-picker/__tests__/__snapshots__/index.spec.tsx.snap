// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ColorPicker > should render correctly 1`] = `
<div>
  <div
    class="
              univer-cursor-default univer-space-y-2 univer-rounded-lg univer-bg-white
              dark:!univer-bg-gray-700
            "
    data-u-comp="color-picker"
  >
    <div
      class="univer-grid univer-content-center univer-gap-2"
      data-u-comp="color-picker-presets"
    >
      <div
        class="univer-grid univer-grid-flow-col univer-items-center univer-justify-between univer-gap-2"
      >
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow !univer-border-gray-200"
          style="background-color: #FFFFFF;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #E1EFFE;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #FDE8E8;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #FEECDC;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #FFF4B9;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #DEF7EC;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #D5F5F6;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #EDEBFE;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #FCE8F3;"
          type="button"
        />
      </div>
      <div
        class="univer-grid univer-grid-flow-col univer-items-center univer-justify-between univer-gap-2"
      >
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #CDD0D8;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #A4CAFE;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #F8B4B4;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #FDBA8C;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #FAC815;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #84E1BC;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #7EDCE2;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #CABFFD;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #F8B4D9;"
          type="button"
        />
      </div>
      <div
        class="univer-grid univer-grid-flow-col univer-items-center univer-justify-between univer-gap-2"
      >
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #979DAC;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #3F83F8;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #F05252;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #FF5A1F;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #D49D0F;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #0DA471;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #0694A2;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #9061F9;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #E74694;"
          type="button"
        />
      </div>
      <div
        class="univer-grid univer-grid-flow-col univer-items-center univer-justify-between univer-gap-2"
      >
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #414657;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #1A56DB;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #C81E1E;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #B43403;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #9A6D15;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #046C4E;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #036672;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #6C2BD9;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #BF125D;"
          type="button"
        />
      </div>
      <div
        class="univer-grid univer-grid-flow-col univer-items-center univer-justify-between univer-gap-2"
      >
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow univer-ring-2 univer-ring-offset-2 univer-ring-offset-white dark:!univer-ring-primary-600 dark:!univer-ring-offset-gray-600"
          style="background-color: #000000;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #233876;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #771D1D;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #771D1D;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #634312;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #014737;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #014451;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #4A1D96;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #751A3D;"
          type="button"
        />
      </div>
    </div>
    <div
      class="univer-flex univer-h-7 univer-items-center"
    >
      <a
        class="
                      univer-cursor-pointer univer-gap-2 univer-text-sm univer-text-gray-900 univer-transition-opacity
                      dark:!univer-text-white
                      hover:univer-opacity-80
                    "
      />
    </div>
  </div>
</div>
`;

exports[`ColorPicker > should render with value 1`] = `
<div>
  <div
    class="
              univer-cursor-default univer-space-y-2 univer-rounded-lg univer-bg-white
              dark:!univer-bg-gray-700
            "
    data-u-comp="color-picker"
  >
    <div
      class="univer-grid univer-content-center univer-gap-2"
      data-u-comp="color-picker-presets"
    >
      <div
        class="univer-grid univer-grid-flow-col univer-items-center univer-justify-between univer-gap-2"
      >
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow !univer-border-gray-200"
          style="background-color: #FFFFFF;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #E1EFFE;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #FDE8E8;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #FEECDC;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #FFF4B9;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #DEF7EC;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #D5F5F6;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #EDEBFE;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #FCE8F3;"
          type="button"
        />
      </div>
      <div
        class="univer-grid univer-grid-flow-col univer-items-center univer-justify-between univer-gap-2"
      >
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #CDD0D8;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #A4CAFE;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #F8B4B4;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #FDBA8C;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #FAC815;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #84E1BC;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #7EDCE2;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #CABFFD;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #F8B4D9;"
          type="button"
        />
      </div>
      <div
        class="univer-grid univer-grid-flow-col univer-items-center univer-justify-between univer-gap-2"
      >
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #979DAC;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #3F83F8;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #F05252;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #FF5A1F;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #D49D0F;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #0DA471;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #0694A2;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #9061F9;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #E74694;"
          type="button"
        />
      </div>
      <div
        class="univer-grid univer-grid-flow-col univer-items-center univer-justify-between univer-gap-2"
      >
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #414657;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #1A56DB;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #C81E1E;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #B43403;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #9A6D15;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #046C4E;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #036672;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #6C2BD9;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #BF125D;"
          type="button"
        />
      </div>
      <div
        class="univer-grid univer-grid-flow-col univer-items-center univer-justify-between univer-gap-2"
      >
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #000000;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #233876;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #771D1D;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #771D1D;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #634312;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #014737;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #014451;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #4A1D96;"
          type="button"
        />
        <button
          class="univer-box-border univer-h-5 univer-w-5 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow"
          style="background-color: #751A3D;"
          type="button"
        />
      </div>
    </div>
    <div
      class="univer-flex univer-h-7 univer-items-center"
    >
      <a
        class="
                      univer-cursor-pointer univer-gap-2 univer-text-sm univer-text-gray-900 univer-transition-opacity
                      dark:!univer-text-white
                      hover:univer-opacity-80
                    "
      />
    </div>
  </div>
</div>
`;
